File: code_snippets/kotlin/src/main/kotlin/usecases/microservices/vo/SubscriptionObject.kt
package usecases.microservices.vo

import dev.restate.sdk.annotation.Handler
import dev.restate.sdk.annotation.VirtualObject
import dev.restate.sdk.http.vertx.RestateHttpEndpointBuilder
import dev.restate.sdk.kotlin.*
import dev.restate.sdk.lambda.BaseRestateLambdaHandler
import dev.restate.sdk.lambda.RestateLambdaEndpointBuilder
import kotlinx.serialization.Serializable

// <start_here>
// <mark_2>
@VirtualObject
class SubscriptionObject {
  // </mark_2>
  companion object {
    // <mark_1>
    val SUBSCRIPTION = KtStateKey.json<String>("subscription")
    // </mark_1>
  }

  // <mark_2>
  @Handler
  suspend fun add(ctx: ObjectContext, req: SubscriptionRequest) {
    // </mark_2>
    // <mark_1>
    ctx.set(SUBSCRIPTION, "awaiting_payment")
    // </mark_1>

    val paymentId = ctx.random().nextUUID().toString()
    val success = ctx.runBlock { createRecurringPayment(req.creditCard, paymentId) }

    if (!success) {
      // <mark_1>
      ctx.set(SUBSCRIPTION, "payment_failed")
      // </mark_1>
      return
    }

    // <mark_1>
    ctx.set(SUBSCRIPTION, "creating_subscription")
    // </mark_1>
    ctx.runBlock { createSubscription(req.userId, req.subscription) }

    // <mark_1>
    ctx.set(SUBSCRIPTION, "created")
    // </mark_1>
  }
}

// <mark_3>
class MyLambdaHandler : BaseRestateLambdaHandler() {
  override fun register(builder: RestateLambdaEndpointBuilder) {
    builder.bind(SubscriptionObject())
  }
}
// </mark_3>
// <end_here>

@Serializable
data class SubscriptionRequest(
    val userId: String,
    val creditCard: String,
    val subscription: String
)

fun createSubscription(userId: String, subscription: String): String {
  return "SUCCESS"
}

fun createRecurringPayment(creditCard: String, paymentId: String): Boolean {
  return true
}

File: code_snippets/go/usecases/microservices/vo/subscriptionobject.go
package vo

import (
	"github.com/aws/aws-lambda-go/lambda"
	restate "github.com/restatedev/sdk-go"
	"github.com/restatedev/sdk-go/server"
	"log"
)

type SubscriptionRequest struct {
	UserID       string `json:"userId"`
	CreditCard   string `json:"creditCard"`
	Subscription string `json:"subscription"`
}

// <start_here>
type SubscriptionObject struct{}

// <mark_2>
func (SubscriptionObject) Add(ctx restate.ObjectContext, req SubscriptionRequest) error {
	// </mark_2>

	// <mark_1>
	restate.Set(ctx, "subscription", "awaiting_payment")
	// </mark_1>
	paymentId := restate.Rand(ctx).UUID().String()
	success, err := restate.Run(ctx, func(ctx restate.RunContext) (bool, error) {
		return CreateRecurringPayment(req.CreditCard, paymentId)
	})
	if err != nil {
		return err
	}
	if !success {
		// <mark_1>
		restate.Set(ctx, "subscription", "payment_failed")
		// </mark_1>
		return nil
	}

	// <mark_1>
	restate.Set(ctx, "subscription", "creating_subscription")
	// </mark_1>
	if _, err := restate.Run(ctx, func(ctx restate.RunContext) (restate.Void, error) {
		return restate.Void{}, CreateSubscription(req.UserID, req.Subscription)
	}); err != nil {
		return err
	}
	// <mark_1>
	restate.Set(ctx, "subscription", "created")
	// </mark_1>

	return nil
}

// <mark_3>
func main() {
	handler, err := server.NewRestate().
		Bind(restate.Reflect(SubscriptionObject{})).
		Bidirectional(false).
		LambdaHandler()
	if err != nil {
		log.Fatal(err)
	}
	lambda.Start(handler)
}

// </mark_3>
// <end_here>

func CreateSubscription(id string, subscription string) error {
	return nil
}

func CreateRecurringPayment(card string, id string) (bool, error) {
	return true, nil
}
